packages/ckeditor5-find-and-replace/src/findandreplacestate.ts

```diff
@@ -166,4 +166,22 @@ export default class FindAndReplaceState extends /* #__PURE__ */ ObservableMixin
 /**
  * The callback function used to find matches in the document.
  */
-export type FindCallback = ( ( { item, text }: { item: Item; text: string } ) => Array<ResultType> );
+export type FindCallback = ( { item, text }: { item: Item; text: string } ) => FindCallbackResult;
+
+/**
+ * Represents the result of a find callback.
+ */
+export type FindCallbackResult =
+	// Deprecated: Try to avoid returning array of results. Return the `searchText` attribute in the result object instead.
+	| Array<ResultType>
+
+	// The `searchText` attribute is used to determine if the search query has changed.
+	// If returned `searchText` is different than the last search query, the search results
+	// will be invalidated after clicking `find next` button and the search will be re-run.
+	//
+	// It's highly recommended to return the `searchText` attribute in the result object instead of
+	// returning array to ensure the search results are properly invalidated when the search query changes.
+	| {
+		results: Array<ResultType>;
+		searchText: string;
+	};
```

Comment:
I'm not sure if we have any convention, or how to do this exactly with TS, but would it be possible to deprecate `Array<ResultType>` with a proper `@deprecated` annotation? 🤔

Reply:
Also if you could move this comment as a part of official part of the docs so it shows up in the API docs. And I would avoid referring to the UI "after clicking `find next` button and", because this behavior is generic (e.g. when using API directly), so this might be confusing.

Reply:
Docs changed.